### PR TITLE
[shared_preferences] Support v2 android embedder.

### PR DESCRIPTION
**Pull request by cyanglaz**
## Description

Migrate shared_preferences to the new android embedding. This PR does the minimum work to migrate the plugin to the 2.0 embedder. No additional refactoring and tests are added.

Note: IAP migration is depended on landing this PR. 

## Related Issues

https://github.com/flutter/flutter/issues/41848

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/shared_preferences/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java [hunk 1]
```diff
@@ -5,199 +5,36 @@
...
-      stream = new ObjectInputStream(new ByteArrayInputStream(Base64.decode(encodedList, 0)));
-      return (List<String>) stream.readObject();
-    } catch (ClassNotFoundException e) {
-      throw new IOException(e);
-    } finally {
-      if (stream != null) {
-        stream.close();
-      }
-    }
+    SharedPreferencesPlugin plugin = new SharedPreferencesPlugin();
```
Comment (bparrishMines) [inline packages/shared_preferences/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java]: nit again:
```suggestion
    final SharedPreferencesPlugin plugin = new SharedPreferencesPlugin();
```

File: packages/shared_preferences/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java [hunk 2]
```diff
@@ -5,199 +5,36 @@
...
-      }
-
-      @Override
-      protected void onPostExecute(Boolean value) {
-        result.success(value);
-      }
-    }.execute();
+  @Override
+  public void onDetachedFromEngine(@NonNull FlutterPlugin.FlutterPluginBinding binding) {
+    channel.setMethodCallHandler(null);
```
Comment (matthew-carroll) [inline packages/shared_preferences/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java]: Recommendation: Consider defining a `teardownChannel()` for symmetry with `setupChannel()`.

File: packages/shared_preferences/example/android/app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -4,12 +4,19 @@
...
-        <activity android:name=".MainActivity"
+        <activity android:name=".EmbeddingV1Activity"
                   android:launchMode="singleTop"
                   android:theme="@android:style/Theme.Black.NoTitleBar"
                   android:configChanges="orientation|keyboardHidden|keyboard|screenSize|locale|layoutDirection"
                   android:hardwareAccelerated="true"
                   android:windowSoftInputMode="adjustResize">
+        </activity>
+        <activity android:name=".MainActivity"
+            android:launchMode="singleTop"
```
Comment (matthew-carroll) [inline packages/shared_preferences/example/android/app/src/main/AndroidManifest.xml]: I made a comment about `singleTop` in the sensor PR.

**Comment by matthew-carroll** [comment]
LGTM

**Pull request merged**